substrate/client/network/sync/src/lib.rs

```diff
@@ -1521,44 +1496,53 @@ where
 
 		if request.fields == BlockAttributes::JUSTIFICATION {
 			match self.on_block_justification(peer_id, block_response) {
-				Ok(OnBlockJustification::Nothing) => None,
-				Ok(OnBlockJustification::Import { peer, hash, number, justifications }) => {
-					self.import_justifications(peer, hash, number, justifications);
-					None
-				},
+				Ok(OnBlockJustification::Nothing) => OnBlockResponse::Nothing,
+				Ok(OnBlockJustification::Import { peer_id, hash, number, justifications }) =>
+					OnBlockResponse::ImportJustifications(ImportJustificationsAction {
+						peer_id,
+						hash,
+						number,
+						justifications,
+					}),
 				Err(BadPeer(id, repu)) => {
 					self.network_service
 						.disconnect_peer(id, self.block_announce_protocol_name.clone());
 					self.network_service.report_peer(id, repu);
-					None
+					OnBlockResponse::Nothing
 				},
 			}
 		} else {
 			match self.on_block_data(&peer_id, Some(request), block_response) {
-				Ok(OnBlockData::Import(origin, blocks)) => {
-					self.import_blocks(origin, blocks);
-					None
-				},
-				Ok(OnBlockData::Request(peer, req)) => Some((peer, req)),
-				Ok(OnBlockData::Continue) => None,
+				Ok(OnBlockData::Import(action)) => OnBlockResponse::ImportBlocks(action),
+				Ok(OnBlockData::Request(peer_id, request)) =>
+					OnBlockResponse::SendBlockRequest { peer_id, request },
+				Ok(OnBlockData::Continue) => OnBlockResponse::Nothing,
 				Err(BadPeer(id, repu)) => {
 					self.network_service
 						.disconnect_peer(id, self.block_announce_protocol_name.clone());
 					self.network_service.report_peer(id, repu);
```

Comment:
Needs a deeper analysis but I'm not sure if `ChainSync` requires a handle to `NetworkService` anymore. Disconnecting & reporting a peer could be another action.

We could probably get rid of the `NetworkServiceProvider` hack if the handle to `NetworkService` was only passed to `SyncingEngine::run()`.

Reply:
> Disconnecting & reporting a peer could be another action.

It's probably fine as long as these actions will be handled synchornously. Just want to make sure we won't end up with reputation changes being pushed to an async queue again. We had problems with that in the past.

Reply:
In fact, right now they are pushed to an async queue: https://github.com/paritytech/polkadot-sdk/blob/0691c91e1558f235081475bfd60ce71035f272a5/substrate/client/network/sync/src/service/network.rs#L83

But luckily after removing `NetworkServiceProvider` we will be able to update reputations synchronously, because `PeerStore` (formerly `PeerSet`) has a sync implementation now: https://github.com/paritytech/polkadot-sdk/blob/0691c91e1558f235081475bfd60ce71035f272a5/substrate/client/network/src/peer_store.rs#L96